src/librados/librados_cxx_buffer.cc

```diff
@@ -0,0 +1,204 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+/// duplicate ceph::buffer implementation within this unit
+#include "common/buffer.cc"
+#include "librados/librados_cxx.h"
+
+/// symbol versioning helpers for ceph::buffer
+#define CEPH_BUFFER_NS_API(fn)                            \
+  LIBRADOS_CXX_API(_ZN4ceph6buffer, fn)
```

Comment:
I'm a bit unclear on how this expected to work. The C++ standard does not guarantee name mangling conventions besides that required for C linkage. So... where does this leave us with compiling for clang, etc.? If the problem being solved is symbol exporting for C, maybe the way to do it is with a wrapper module?

Reply:
GCC and Clang both follow the IA64 C++ mangling conventions [1]. For environments that don't support it, the fallback would be to just export everything w/o symbol versioning.


[1] https://itanium-cxx-abi.github.io/cxx-abi/

Reply:
correct. C++ standard does not define the name mangling convention. but in practice, the mainstream compilers like GCC, Clang, Intel C++ compiler all use the IA64 C++ ABI Spec. see https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling, which is well defined.

Reply:
and see https://github.com/gcc-mirror/gcc/blob/master/gcc/cp/mangle.c

Reply:
Agreed that the convention is relatively stable:
https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html

Note that name mangling can change across architectures-- when sizeof(int) changes. GCC for instance has changed before. Is it possible to accomplish this with inline namespaces in our case? If not, ok, at least we're doing it with open eyes! :-)

Reply:
Hmm -- that inline namespace idea actually sounds much cleaner. Thanks.

Reply:
... with the apparent side-effect that all the existing forward declarations to librados C++ classes need to be yanked or wrapped with a version-specific inline. 

e.g. lots of code does things like "namespace librados { class AioCompletion; }" which make AioCompletion ambiguous when its moved to an inline namespace under librados.

Reply:
If I'm following your meaning, then yeah there would have to be some revision. I've used the feature in a basic context:
namespace foo { inline namespace version1 { ... } } 
But I haven't had opportunity to really consider the cases of reopening namespaces and understanding how that interacts with this feature yet. Looks like we might get to! Hope that helps.

Reply:
I added a librados forward declarations header file and yanked the individual forward declarations throughout. I think this inline namespace technique is a much nicer path forward.